### PR TITLE
Merge pipelines generate-releasenotes.yml into version-updater.yml

### PR DESCRIPTION
**Pull request by chidozieononiwu**
- Merging Pipelines
- Add only packages that were recently updated.

### CODE REVIEW

File: eng/pipelines/template/generate-releasenotes.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+parameters:
```
Comment (weshaggard) [inline eng/pipelines/template/generate-releasenotes.yml:1]: If you want to follow the convention this file should also be under a steps directory.

File: eng/scripts/Update-Release-Versions.ps1 [hunk 2]
```diff
@@ -209,6 +211,10 @@ function Update-Packages($lang, $packageList, $langVersions, $langLinkTemplates)
       if (CheckRequiredLinks $langLinkTemplates $pkg $version){
         Write-Host "Updating VersionGA $($pkg.Package) from $($pkg.VersionGA) to $version"
         $pkg.VersionGA = $version;
+        Write-Host "[debug]$pkg"
+        $updatedGAPackage = $pkg.PSObject.Copy()
```
Comment (weshaggard) [inline eng/scripts/Update-Release-Versions.ps1:215]: Is the copy needed because you are adding a new property?

Why is the new property interesting because you don't know which one of GA or Preview was updated?

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 3]
```diff
@@ -1,84 +1,119 @@
...
-Write-Host "Common Script $commonScript"
+#. $commonScript
+#$CsvMetaData = Get-CSVMetadata -MetadataUri "https://raw.githubusercontent.com/azure-sdk/azure-sdk/PackageVersionUpdates/_data/releases/latest/${releaseFileName}-packages.csv"
 
-. $commonScript
-$CsvMetaData = Get-CSVMetadata -MetadataUri "https://raw.githubusercontent.com/azure-sdk/azure-sdk/PackageVersionUpdates/_data/releases/latest/${releaseFileName}-packages.csv"
-
-$releaseFilePath = (Join-Path $ReleaseDirectory $releasePeriod "${releaseFileName}.md")
-$pathToRelatedYaml = (Join-Path $ReleaseDirectory ".." _data releases $releasePeriod "${releaseFileName}.yml")
+$releaseFilePath = (Join-Path $ReleaseDirectory $releasePeriod "${repoLanguage}.md")
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:115]: We shouldn't need the releasefilePath any more.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 4]
```diff
@@ -1,84 +1,119 @@
...
-  exit 1
-}
-Write-Host "Repo Path $LanguageRepoDirectory"
+. (Join-Path $commonScriptPath ChangeLog-Operations.ps1)
+. (Join-Path $commonScriptPath SemVer.ps1)
 
-$commonScriptsPath = (Join-Path $LanguageRepoDirectory eng common scripts)
-$commonScript = (Join-Path $commonScriptsPath common.ps1)
-Write-Host "Common Script $commonScript"
+#. $commonScript
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:112]: I assume you plan to clean these up.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 5]
```diff
@@ -1,84 +1,119 @@
...
+  [string]$changedPackagesPath
 )
 
-# Filters out packages if the versions are not found in the packages.csv file.
-# Also appends DisplayName and ServiceName to the releaseHighlights
-function FilterOut-UnreleasedPackages ($releaseHighlights)
+function AddChangelogUrls ()
 {
-    $results = @()
+    $changedPackages = (Get-Content -Path $changedPackagesPath | ConvertFrom-Json).where({ $_.Language -eq $repoLanguage })
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:11]: I suggest making either the path or the package set a parameter to the function instead of using variables from the global scope.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 6]
```diff
@@ -1,84 +1,119 @@
...
+            $language = "net"
+            $package.Language = $language
+        }
+        $packageName = $package.Package
+        $packageDirName = $package.Package
+        if ($language -eq "js")
         {
-            $sortedVersions = [AzureEngSemanticVersion]::SortVersionStrings(@($releaseVersion, $packageMetaData.VersionGA, $packageMetaData.VersionPreview))
+            # Remove @azure/
+            $packageDirName = ($packageDirName.split("/"))[1]
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:30]: Lets try to avoid doing this duplicate processing in this script and in the update-release-versions script. Instead lets use our link processing in update-release-versions to build a source link that we can pass to this script.

File: eng/scripts/Update-Release-Versions.ps1 [hunk 7]
```diff
@@ -230,13 +236,18 @@ function Update-Packages($lang, $packageList, $langVersions, $langLinkTemplates)
       if (CheckRequiredlinks $langLinkTemplates $pkg $version) {
         Write-Host "Updating VersionPreview $($pkg.Package) from $($pkg.VersionPreview) to $version"
         $pkg.VersionPreview = $version;
+        Write-Host "[debug]$pkg"
+        $updatedPreviewPackage = $pkg.PSObject.Copy()
+        $updatedPreviewPackage | Add-Member -NotePropertyName "UpdatedVersion" -NotePropertyValue $version
```
Comment (weshaggard) [inline eng/scripts/Update-Release-Versions.ps1:241]: Lets as add a SourceLink property here as we can get it similar to how CheckRequiredLinks function does and we won't need to repeat that logic in both scripts.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 8]
```diff
@@ -1,84 +1,119 @@
...
+        }
+        LogDebug "language: $language"
+        $package | Add-Member -NotePropertyName "PackageDirName" -NotePropertyValue $packageDirName
+        $serviceName = $package.RepoPath
+        $updatedVersion = $package.UpdatedVersion
+        $changelogRawUrl = "https://raw.githubusercontent.com/Azure/azure-sdk-for-" + `
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:36]: This logic will break in some cases so as I've mentioned in other comments lets not do it here but instead only in update-release-versions.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 9]
```diff
@@ -1,84 +1,119 @@
...
+            $changelogContent = Invoke-RestMethod -Method GET -Uri $package.UpdatedChangelogUrl -MaximumRetryCount 2
+        }
+        catch
+        {
+            # Skip if the changelog Url is invalid
+            LogWarning "Failed to get content from $($package.UpdatedChangelogUrl)"
+            continue
+        }
+        $tempFile = New-TemporaryFile
+        Set-Content -Path $tempFile.FullName -Value $changelogContent
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:60]: Can we avoid writing these to disk? Can we make a version of the Get-ChangeLogEntries that simply takes the file content instead?

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 10]
```diff
@@ -1,84 +1,119 @@
...
+        $changeLogEntries = Get-ChangeLogEntries -ChangeLogLocation $tempFile.FullName
+        foreach ($changelogEntry in $changeLogEntries.Values)
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:62]: Why do we need to loop? Isn't changelogEntries a dictionary?

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 11]
```diff
@@ -1,84 +1,119 @@
...
+            if ($changeLogEntry.ReleaseVersion -ne $package.UpdatedVersion)
             {
+                LogDebug "Skipping States: $($changeLogEntry.ReleaseVersion) released on $($changeLogEntry.UpdatedVersion)"
                 continue
-                LogDebug "[ $packageName ] with version  [$releaseVersion ] was skipped because it has a newer release version than present in the package csv"
             }
-            $packageSemVer = [AzureEngSemanticVersion]::ParseVersionString($releaseVersion)
-            
+
+            $githubAnchor = $changeLogEntry.ReleaseTitle.Replace("## ", "").Replace(".", "").Replace("(", "").Replace(")", "").Replace(" ", "-")
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:70]: Do we really care about anchors given that we are linking to the changelog via release tag and the top entry should be the correct entry?

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 12]
```diff
@@ -1,84 +1,119 @@
...
+
+            $releaseTag = "${package.Package}_${$package.UpdatedVersion}"
+            $ChangelogUrl = "https://github.com/Azure/azure-sdk-for-" + `
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:73]: Why are we building the link again? Couldn't we just replace the raw.githubusercontent.com or even make the url property be relative where we can add the correct root in the right context.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 13]
```diff
@@ -1,84 +1,119 @@
...
+                            "$($package.Language)/blob/${releaseTag}/sdk/$($package.RepoPath)/$($package.PackageDirName)/CHANGELOG.md#${githubAnchor}"
+            $packageSemVer = [AzureEngSemanticVersion]::ParseVersionString($package.UpdatedVersion)
+            $releaseEntryContent = @()
+            $changeLogEntry.ReleaseContent | %{
+                $releaseEntryContent += $_.Replace("###", "####")
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:78]: Why is this necessary?

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 14]
```diff
@@ -1,84 +1,119 @@
...
+                ChangelogContent = ($releaseEntryContent | Out-String).Trim()
             }
 
             if (!$entry.DisplayName)
             {
                 $entry.DisplayName = $entry.Name
             }
 
-            if ($packageGroupId)
+            if ($package.GroupId)
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:97]: In strict mode this will fail for everything but java. You might need to tweak it to handle it by checking for the property similar to https://github.com/Azure/azure-sdk/blob/master/eng/scripts/PackageList-Helpers.ps1#L147.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 15]
```diff
@@ -95,37 +130,43 @@ Register-PSRepository -Name azure-sdk-tools-feed -SourceLocation $ToolsFeed -Pub
...
+            continue
         }
+        $filteredEntries += $entry
     }
 }
-
-$filteredReleaseHighlights = FilterOut-UnreleasedPackages -releaseHighlights $incomingReleaseHighlights
+else 
+{
+    $filteredEntries = $incomingReleaseEntries
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:152]:
```suggestion
    $existingYamlContent.entries= $incomingReleaseEntries
```

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 16]
```diff
@@ -95,37 +130,43 @@ Register-PSRepository -Name azure-sdk-tools-feed -SourceLocation $ToolsFeed -Pub
...
+    foreach ($entry in $incomingReleaseEntries)
     {
-        $presentKey = $existingYamlContent.entries.Where( { ($_.name + ':' + $_.version) -eq $key } )
+        $presentKey = $existingYamlContent.entries.Where( { ($_.name -eq $entry.name) -and ($_.version -eq $entry.version) } )
         if ($presentKey.Count -gt 0)
         {
-            $incomingReleaseHighlights.Remove($key)
+            continue
         }
+        $filteredEntries += $entry
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:147]:
```suggestion
        $existingYamlContent.entries += $entry
```

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 17]
```diff
@@ -95,37 +130,43 @@ Register-PSRepository -Name azure-sdk-tools-feed -SourceLocation $ToolsFeed -Pub
...
+    $filteredEntries = $incomingReleaseEntries
+}
 
 if ($null -eq $existingYamlContent.entries)
 {
     $existingYamlContent.entries = New-Object "System.Collections.Generic.List[System.Collections.Specialized.OrderedDictionary]"
 }
 
-foreach ($entry in $filteredReleaseHighlights)
+foreach ($entry in $filteredEntries)
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:160]: We shouldn't need this other loop we can do it inline with the suggestions I provided above.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 18]
```diff
@@ -95,37 +130,43 @@ Register-PSRepository -Name azure-sdk-tools-feed -SourceLocation $ToolsFeed -Pub
...
+}
 
-if (!(Test-Path $releaseFilePath) -and $existingYamlContent.entries.Count -gt 0)
-{
-    &(Join-Path $PSScriptRoot Generate-Release-Structure.ps1) -releaseFileName "${releaseFileName}.md" -ExcludeFileNames @()
-}
\ No newline at end of file
+#Set-Content -Path $pathToRelatedYaml -Value (ConvertTo-Yaml $existingYamlContent)
+#
+#if (!(Test-Path $releaseFilePath) -and $existingYamlContent.entries.Count -gt 0)
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:169]: I already removed these so you will need to rebase on my changes.

File: eng/scripts/Update-Release-Versions.ps1 (hunk 2 above)
Comment (chidozieononiwu) [reply eng/scripts/Update-Release-Versions.ps1:228]: That's right. Need to know if it GA or Preview

File: _includes/releases/variables/android.md [hunk 19]
```diff
@@ -5,4 +5,6 @@
 {% assign package_url_template = "https://search.maven.org/artifact/item.GroupId/item.Package/item.Version/aar" %}
 <!--{% assign msdocs_url_template = "https://docs.microsoft.com/android/api/overview/azure/item.TrimmedPackage-readme" %}-->
 <!--{% assign ghdocs_url_template = "https://azuresdkdocs.blob.core.windows.net/$web/android/item.Package/item.Version/index.html" %}-->
-{% assign source_url_template = "https://github.com/Azure/azure-sdk-for-android/tree/item.Package_item.Version/sdk/item.RepoPath/item.Package" %}
\ No newline at end of file
+{% assign source_url_template = "https://github.com/Azure/azure-sdk-for-android/tree/item.Package_item.Version/sdk/item.RepoPath/item.Package" %}
```
Comment (weshaggard) [inline _includes/releases/variables/android.md:8]: Do we really need the 2 new templates? I was expecting we could just use the source_url_template and add the CHANGELOG.md and if we needed the content we could replace the root domain.

File: eng/scripts/Update-Release-Versions.ps1 [hunk 20]
```diff
@@ -132,8 +134,19 @@ function CheckRequiredLinks($linkTemplates, $pkg, $version)
   }
 
   $pkgLink = GetTemplateValue $linkTemplates "package_url_template" $pkg.Package $version $pkg.RepoPath $groupId
-
+  Write-Host "[debug]PkgLink $pkgLink"
   $valid = $valid -and (CheckLink $pkgLink)
+
+  $changelogLink = GetTemplateValue $linkTemplates "changelog_blob_url_template" $pkg.Package $version $pkg.RepoPath
```
Comment (weshaggard) [inline eng/scripts/Update-Release-Versions.ps1:140]: Can we instead just build this from the source link? The reason I say that is because there are some cases were the template is completely overwritten because the RepoPath contains a full url instead of service directory. See link 124 above.

File: eng/scripts/Update-Release-Versions.ps1 [hunk 21]
```diff
@@ -209,6 +223,11 @@ function Update-Packages($lang, $packageList, $langVersions, $langLinkTemplates)
       if (CheckRequiredLinks $langLinkTemplates $pkg $version){
         Write-Host "Updating VersionGA $($pkg.Package) from $($pkg.VersionGA) to $version"
         $pkg.VersionGA = $version;
+        Write-Host "[debug]$pkg"
+        Write-Host "[debug]LanguageLinkTemplates $($langLinkTemplates.Values)"
+        $updatedGAPackage = $pkg.PSObject.Copy()
+        $updatedGAPackage | Add-Member -NotePropertyName "UpdatedVersion" -NotePropertyValue $version
```
Comment (weshaggard) [inline eng/scripts/Update-Release-Versions.ps1:229]: I was expecting to see you add a ChangeLogUrl link as well but I don't see you adding that property yet.

File: eng/scripts/Generate-ReleaseNotes.ps1 (hunk 13 above)
Comment (chidozieononiwu) [reply eng/scripts/Generate-ReleaseNotes.ps1:78]: This is because in the combined release content for each entry starts with a level 3 header. So this is to make sure there is not two level 3 headers. Although I think the right thing here should be to bump all subsequent headers by 1.

File: _includes/releases/variables/android.md (hunk 19 above)
Comment (chidozieononiwu) [reply _includes/releases/variables/android.md:8]: The source_url_template used "tree" rather than "blob". That might work for the regular Changelog URI but it won't work for the URI to the raw file. I also though it is better to add the template for both rather than trying to split and recreate the URL later.

File: eng/scripts/Update-Release-Versions.ps1 (hunk 20 above)
Comment (chidozieononiwu) [reply eng/scripts/Update-Release-Versions.ps1:140]: Am handling those cases by logging an error if the call to get the content of the URI failed later on then skipping. I can use the `source_url_template` here but it still won't work for the raw URI link.

File: eng/scripts/Generate-ReleaseNotes.ps1 (hunk 17 above)
Comment (chidozieononiwu) [reply eng/scripts/Generate-ReleaseNotes.ps1:157]: Unfortunately the Yaml parser will fail if $entries is $null and not initialized to "System.Collections.Generic.List[System.Collections.Specialized.OrderedDictionary]"

File: eng/scripts/Update-Release-Versions.ps1 [hunk 22]
```diff
@@ -209,6 +211,20 @@ function Update-Packages($lang, $packageList, $langVersions, $langLinkTemplates)
       if (CheckRequiredLinks $langLinkTemplates $pkg $version){
         Write-Host "Updating VersionGA $($pkg.Package) from $($pkg.VersionGA) to $version"
         $pkg.VersionGA = $version;
+        Write-Host "[debug]$pkg"
+
+        $changelogBlobLink = GetTemplateValue $langLinkTemplates "changelog_blob_url_template" $pkg.Package $version $pkg.RepoPath
```
Comment (weshaggard) [inline eng/scripts/Update-Release-Versions.ps1:216]:
```suggestion
        $changelogBlobLink = (GetTemplateValue $langLinkTemplates "source_url_template " $pkg.Package $version $pkg.RepoPath) + "/CHANGELOG.md"
        $changelogRawLink = $changelogBlobLink -replace "https://github.com/(.*)/(tree|blob)", "https://raw.githubusercontent.com/`$1"
```

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 23]
```diff
@@ -1,85 +1,75 @@
...
         {
-            $sortedVersions = [AzureEngSemanticVersion]::SortVersionStrings(@($releaseVersion, $packageMetaData.VersionGA, $packageMetaData.VersionPreview))
-
-            if (($sortedVersions[0] -eq $releaseVersion) -and ($releaseVersion -ne $packageMetaData.VersionGA) -and ($releaseVersion -ne $packageMetaData.VersionPreview))
-            {
-                continue
-                LogDebug "[ $packageName ] with version  [$releaseVersion ] was skipped because it has a newer release version than present in the package csv"
+            $packageSemVer = [AzureEngSemanticVersion]::ParseVersionString($package.UpdatedVersion)
+            $releaseEntryContent = @()
+            $changeLogEntry.ReleaseContent | %{
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:33]: Lets do this processing in the jekyll template file instead of this automation so that we can ensure it matches what is going to be displayed.

File: eng/pipelines/version-updater.yml [hunk 24]
```diff
@@ -7,7 +7,10 @@ resources:
       type: github
       name: Azure/azure-sdk-tools
       endpoint: azure
-      ref: refs/tags/azure-sdk-tools_20210422.2
+#      ref: refs/tags/azure-sdk-tools_20210422.2
```
Comment (weshaggard) [inline eng/pipelines/version-updater.yml:10]: Lets make sure to uncomment this once a new tag version is pulled in.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 25]
```diff
@@ -95,37 +87,30 @@ Register-PSRepository -Name azure-sdk-tools-feed -SourceLocation $ToolsFeed -Pub
...
 {
+    $filteredEntries = $incomingReleaseEntries
     $existingYamlContent.entries = New-Object "System.Collections.Generic.List[System.Collections.Specialized.OrderedDictionary]"
 }
 
-foreach ($entry in $filteredReleaseHighlights)
-{
-    $existingYamlContent.entries += $entry
-}
+ForEach-Object -InputObject $filteredEntries -Process {$existingYamlContent.entries += $_}
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:113]: Is this needed here still? If so I'd prefer the standard foreach loop however I still think you can do it inline above.

File: eng/scripts/Generate-ReleaseNotes.ps1 (hunk 23 above)
Comment (chidozieononiwu) [reply eng/scripts/Generate-ReleaseNotes.ps1:35]: I think it's still better to do this here. Trying to modify the lines in Jekyll requires splitting the string into an array or trying to replace across the multiline string. The Jekyll syntax does not seem to be advanced enough for any of this.  Splitting it into array causes a whole new range of issues for the rendering especially for code-block content.

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 26]
```diff
@@ -93,32 +95,29 @@ Register-PSRepository -Name azure-sdk-tools-feed -SourceLocation $ToolsFeed -Pub
...
 }
-
-$filteredReleaseHighlights = FilterOut-UnreleasedPackages -releaseHighlights $incomingReleaseHighlights
-
-if ($null -eq $existingYamlContent.entries)
+else 
 {
-    $existingYamlContent.entries = New-Object "System.Collections.Generic.List[System.Collections.Specialized.OrderedDictionary]"
+    $filteredEntries = $incomingReleaseEntries
+    $existingYamlContent.entries = $filteredEntries
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:118]: Based on your logic shouldn't this need to be outside of the if/else so that both blocks assign this?

File: eng/scripts/Generate-ReleaseNotes.ps1 (hunk 23 above)
Comment (weshaggard) [reply eng/scripts/Generate-ReleaseNotes.ps1]: Ok we can keep this here for now but I'd like to reconsider other options in the future.

File: eng/pipelines/template/steps/generate-releasenotes.yml [hunk 27]
```diff
@@ -62,17 +47,23 @@ steps:
...
         PROwner: Azure
+        BaseBranchName: ${{ parameters.DefaultBranch }}
         PRBranchName: '${{ parameters.PrBranchName }}_${{ repo.value.Language }}_$(ReleasePeriod)'
         CommitMsg: "${{ repo.value.Language }} release notes for the $(ReleasePeriod) release"
         PRTitle: "${{ repo.value.Language }} release notes for the $(ReleasePeriod) release"
         PRBody: "This is an automated pull request to gather the ${{ repo.value.Language }} release notes for the $(ReleasePeriod) release. See [Producing Release Notes](https://github.com/Azure/azure-sdk/blob/master/docs/policies/releasenotes.md#producing-release-notes) for details of the processing."
         WorkingDirectory: $(System.DefaultWorkingDirectory)/azure-sdk
         ScriptDirectory: $(System.DefaultWorkingDirectory)/azure-sdk-tools/eng/common/scripts
+
+  - pwsh: |
```
Comment (weshaggard) [inline eng/pipelines/template/steps/generate-releasenotes.yml:67]: Why is this needed?

File: eng/scripts/Generate-ReleaseNotes.ps1 [hunk 28]
```diff
@@ -93,32 +95,28 @@ Register-PSRepository -Name azure-sdk-tools-feed -SourceLocation $ToolsFeed -Pub
...
     }
 }
-
-$filteredReleaseHighlights = FilterOut-UnreleasedPackages -releaseHighlights $incomingReleaseHighlights
-
-if ($null -eq $existingYamlContent.entries)
+else 
 {
-    $existingYamlContent.entries = New-Object "System.Collections.Generic.List[System.Collections.Specialized.OrderedDictionary]"
+    $filteredEntries = $incomingReleaseEntries
```
Comment (weshaggard) [inline eng/scripts/Generate-ReleaseNotes.ps1:117]: Does this work? won't it override the type of filteredEntries? I thought that was why you tried to add things individually. If it does end up changing the type you can consider using AddRange.

**Review by weshaggard — APPROVED** [review]
A few comments but it looks ready. After you merge it please keep an eye on the runs to ensure it is doing what is expected.

File: eng/pipelines/template/steps/generate-releasenotes.yml (hunk 27 above)
Comment (chidozieononiwu) [reply eng/pipelines/template/steps/generate-releasenotes.yml:67]: Just resetting the branch before going ahead with other steps in the pipeline.

File: eng/scripts/Generate-ReleaseNotes.ps1 (hunk 28 above)
Comment (chidozieononiwu) [reply eng/scripts/Generate-ReleaseNotes.ps1:117]: I also updated the type returned by the `GetReleaseNotesData` function.
